File: src/main/java/sharev/domain/member/service/MemberService.java
package sharev.domain.member.service;

import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sharev.domain.account.entity.Account;
import sharev.domain.account.repository.AccountRepository;
import sharev.domain.member.dto.response.ResponseMemberDto;
import sharev.domain.member.entity.Member;
import sharev.domain.member.entity.MemberRoleType;
import sharev.domain.member.entity.MemberStatusType;
import sharev.domain.member.repository.MemberRepository;
import sharev.domain.team.entity.Team;
import sharev.domain.team.repository.TeamRepository;
import sharev.exception.CustomException;
import sharev.exception.ExceptionCode;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MemberService {
    private final TeamRepository teamRepository;
    private final MemberRepository memberRepository;
    private final AccountRepository accountRepository;

    public boolean isAdmin(Account account, Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Optional<Member> optionalMember = memberRepository.findByTeamAndAccount(team, account);

        if (optionalMember.isEmpty()) {
            throw new CustomException(ExceptionCode.TEAM_NOT_FOUND);
        }

        Member member = optionalMember.get();

        return member.getRole() == MemberRoleType.ADMIN;
    }

    public List<ResponseMemberDto> getMembers(Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        return memberRepository.findAllByTeam(team).stream()
                .map(ResponseMemberDto::new)
                .toList();
    }

    @Transactional
    public void invite(Account account, Long teamId, String email) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Account targetAccount = accountRepository.findByEmail(email)
                .orElseThrow(() -> new CustomException(ExceptionCode.ACCOUNT_NOT_FOUND));

        memberRepository.findByTeamAndAccount(team, targetAccount)
                .ifPresent(m -> {
                    throw new CustomException(ExceptionCode.MEMBER_ALREADY_EXISTS);
                });

        memberRepository.save(new Member(team, targetAccount, MemberStatusType.INVITE, MemberRoleType.COMMON));
    }

    @Transactional
    public void acceptInvitation(Account account, Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Member member = memberRepository.findByTeamAndAccount(team, account)
                .orElseThrow(() -> new CustomException(ExceptionCode.MEMBER_NOT_FOUND));

        if (member.getStatus() != MemberStatusType.INVITE) {
            throw new CustomException(ExceptionCode.MEMBER_NOT_INVITED);
        }

        member.activate();
    }

    @Transactional
    public void leave(Account account, Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Member member = memberRepository.findByTeamAndAccount(team, account)
                .orElseThrow(() -> new CustomException(ExceptionCode.MEMBER_NOT_FOUND));

        validateNotLastAdmin(team, member);

        memberRepository.delete(member);
    }

    @Transactional
    public void updateRole(Long teamId, Long memberId, MemberRoleType role) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Member member = findMemberInTeam(memberId, team);

        if (role != MemberRoleType.ADMIN) {
            validateNotLastAdmin(team, member);
        }

        member.updateRole(role);
    }

    @Transactional
    public void removeMember(Account account, Long teamId, Long memberId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ExceptionCode.TEAM_NOT_FOUND));

        Member member = findMemberInTeam(memberId, team);

        if (member.getAccount().getId().equals(account.getId())) {
            throw new CustomException(ExceptionCode.CANNOT_REMOVE_SELF);
        }

        validateNotLastAdmin(team, member);

        memberRepository.delete(member);
    }

    private Member findMemberInTeam(Long memberId, Team team) {
        Member member = memberRepository.findById(memberId)
                .orElseThrow(() -> new CustomException(ExceptionCode.MEMBER_NOT_FOUND));

        if (!member.getTeam().getId().equals(team.getId())) {
            throw new CustomException(ExceptionCode.MEMBER_NOT_FOUND);
        }

        return member;
    }

    private void validateNotLastAdmin(Team team, Member member) {
        if (member.getRole() == MemberRoleType.ADMIN
                && memberRepository.countByTeamAndRole(team, MemberRoleType.ADMIN) <= 1) {
            throw new CustomException(ExceptionCode.CANNOT_REMOVE_LAST_ADMIN);
        }
    }
}

File: src/main/kotlin/sharev/member/application/service/MemberService.kt
package sharev.member.application.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import sharev.member.application.port.inbound.command.*
import sharev.member.application.port.inbound.mapper.*
import sharev.member.application.port.inbound.result.*
import sharev.member.application.port.inbound.usecase.*
import sharev.member.application.port.outbound.DeleteMemberPort
import sharev.member.application.port.outbound.LoadAccountForMemberPort
import sharev.member.application.port.outbound.LoadMemberPort
import sharev.member.application.port.outbound.SaveMemberPort
import sharev.member.domain.exception.MemberException
import sharev.member.domain.exception.MemberExceptionCode
import sharev.member.domain.model.Member
import sharev.member.domain.model.MemberRole
import sharev.member.domain.model.MemberStatus
import sharev.team.application.port.outbound.CheckTeamMemberPort
import sharev.team.application.port.outbound.LoadTeamPort
import sharev.team.domain.exception.TeamException
import sharev.team.domain.exception.TeamExceptionCode

@Service
@Transactional(readOnly = true)
class MemberService(
    private val loadTeamPort: LoadTeamPort,
    private val loadMemberPort: LoadMemberPort,
    private val saveMemberPort: SaveMemberPort,
    private val deleteMemberPort: DeleteMemberPort,
    private val loadAccountForMemberPort: LoadAccountForMemberPort,
    private val checkTeamMemberPort: CheckTeamMemberPort,
) : GetMembersUseCase,
    InviteMemberUseCase,
    AcceptInvitationUseCase,
    LeaveTeamUseCase,
    UpdateMemberRoleUseCase,
    RemoveMemberUseCase,
    CheckTeamAdminUseCase {

    override fun isAdmin(teamId: Long, accountId: Long): Boolean {
        if (!loadTeamPort.exists(teamId)) {
            return false
        }

        return checkTeamMemberPort.isAdminMember(accountId, teamId)
    }

    override fun getMembers(command: GetMembersCommand): List<MemberResult> {
        validateTeamMember(command.accountId, command.teamId)

        return loadMemberPort.loadAllByTeam(command.teamId)
            .map { it.toResult() }
    }

    @Transactional
    override fun invite(command: InviteMemberCommand): InviteMemberResult {
        validateTeamAdmin(command.accountId, command.teamId)

        val targetAccountId = loadAccountForMemberPort.loadAccountIdByEmail(command.email)

        if (checkTeamMemberPort.isMember(targetAccountId, command.teamId)) {
            throw MemberException(MemberExceptionCode.MEMBER_ALREADY_EXISTS)
        }

        return saveMemberPort.save(command.teamId, targetAccountId, MemberStatus.INVITE, MemberRole.COMMON)
            .toInviteMemberResult()
    }

    @Transactional
    override fun acceptInvitation(command: AcceptInvitationCommand): AcceptInvitationResult {
        val member = loadMemberPort.loadByTeamAndAccount(command.teamId, command.accountId)

        if (member.status != MemberStatus.INVITE) {
            throw MemberException(MemberExceptionCode.MEMBER_NOT_INVITED)
        }

        return saveMemberPort.activate(member.id)
            .toAcceptInvitationResult()
    }

    @Transactional
    override fun leave(command: LeaveTeamCommand): LeaveTeamResult {
        val member = loadMemberPort.loadByTeamAndAccount(command.teamId, command.accountId)
        validateNotLastAdmin(member)
        deleteMemberPort.delete(member.id)

        return member.toLeaveTeamResult()
    }

    @Transactional
    override fun updateRole(command: UpdateMemberRoleCommand): UpdateMemberRoleResult {
        validateTeamAdmin(command.accountId, command.teamId)

        val member = findMemberInTeam(command.memberId, command.teamId)

        if (command.role != MemberRole.ADMIN) {
            validateNotLastAdmin(member)
        }

        return saveMemberPort.updateRole(member.id, command.role)
            .toUpdateMemberRoleResult()
    }

    @Transactional
    override fun removeMember(command: RemoveMemberCommand): RemoveMemberResult {
        validateTeamAdmin(command.accountId, command.teamId)

        val member = findMemberInTeam(command.memberId, command.teamId)

        if (member.accountId == command.accountId) {
            throw MemberException(MemberExceptionCode.CANNOT_REMOVE_SELF)
        }

        validateNotLastAdmin(member)
        deleteMemberPort.delete(member.id)

        return member.toRemoveMemberResult()
    }

    private fun findMemberInTeam(memberId: Long, teamId: Long): Member {
        val member = loadMemberPort.load(memberId)

        if (member.teamId != teamId) {
            throw MemberException(MemberExceptionCode.MEMBER_NOT_FOUND)
        }

        return member
    }

    private fun validateTeamMember(accountId: Long, teamId: Long) {
        if (!checkTeamMemberPort.isMember(accountId, teamId)) {
            throw TeamException(TeamExceptionCode.NOT_TEAM_MEMBER)
        }
    }

    private fun validateTeamAdmin(accountId: Long, teamId: Long) {
        if (!checkTeamMemberPort.isAdminMember(accountId, teamId)) {
            throw TeamException(TeamExceptionCode.NOT_TEAM_ADMIN_MEMBER)
        }
    }

    private fun validateNotLastAdmin(member: Member) {
        if (member.role == MemberRole.ADMIN &&
            loadMemberPort.countByTeamAndRole(member.teamId, MemberRole.ADMIN) <= 1
        ) {
            throw MemberException(MemberExceptionCode.CANNOT_REMOVE_LAST_ADMIN)
        }
    }
}
